// openvino-genai-node/dist/utils.js
// Copyright (C) 2023-2026 Intel Corporation
// SPDX-License-Identifier: Apache-2.0
export var StreamingStatus;
(function (StreamingStatus) {
    StreamingStatus[StreamingStatus["RUNNING"] = 0] = "RUNNING";
    StreamingStatus[StreamingStatus["STOP"] = 1] = "STOP";
    StreamingStatus[StreamingStatus["CANCEL"] = 2] = "CANCEL";
})(StreamingStatus || (StreamingStatus = {}));
/** controls the stopping condition for grouped beam search. The following values are possible?:
 - "EARLY" stops as soon as there are `num_beams` complete candidates.
 - "HEURISTIC" stops when is it unlikely to find better candidates.
 - "NEVER" stops when there cannot be better candidates.
 */
export var StopCriteria;
(function (StopCriteria) {
    StopCriteria[StopCriteria["EARLY"] = 0] = "EARLY";
    StopCriteria[StopCriteria["HEURISTIC"] = 1] = "HEURISTIC";
    StopCriteria[StopCriteria["NEVER"] = 2] = "NEVER";
})(StopCriteria || (StopCriteria = {}));
/** This object is used to store the configuration for structured generation, which includes
 * the JSON schema and other related parameters. */
export class StructuredOutputConfig {
    constructor(params) {
        const { json_schema, regex, grammar, structural_tags_config } = params;
        this.json_schema = json_schema;
        this.regex = regex;
        this.grammar = grammar;
        this.structural_tags_config = structural_tags_config;
    }
    static JSONSchema(value) {
        return { structuralTagType: "JSONSchema", value };
    }
    static Regex(value) {
        return { structuralTagType: "Regex", value };
    }
    static EBNF(value) {
        return { structuralTagType: "EBNF", value };
    }
    static ConstString(value) {
        return { structuralTagType: "ConstString", value };
    }
    static AnyText() {
        return { structuralTagType: "AnyText" };
    }
    static QwenXMLParametersFormat(jsonSchema) {
        return { structuralTagType: "QwenXMLParametersFormat", jsonSchema };
    }
    static Concat(...elements) {
        return { structuralTagType: "Concat", elements };
    }
    static Union(...elements) {
        return { structuralTagType: "Union", elements };
    }
    static Tag(params) {
        return Object.assign({ structuralTagType: "Tag" }, params);
    }
    static TriggeredTags(params) {
        var _a, _b;
        return Object.assign(Object.assign({ structuralTagType: "TriggeredTags" }, params), { atLeastOne: (_a = params.atLeastOne) !== null && _a !== void 0 ? _a : false, stopAfterFirst: (_b = params.stopAfterFirst) !== null && _b !== void 0 ? _b : false });
    }
    static TagsWithSeparator(params) {
        var _a, _b;
        return Object.assign(Object.assign({ structuralTagType: "TagsWithSeparator" }, params), { atLeastOne: (_a = params.atLeastOne) !== null && _a !== void 0 ? _a : false, stopAfterFirst: (_b = params.stopAfterFirst) !== null && _b !== void 0 ? _b : false });
    }
}
